File: archivist/src/test/java/com/zorroa/archivist/repository/DyHierarchyDaoTests.java
package com.zorroa.archivist.repository;

import com.google.common.collect.ImmutableList;
import com.zorroa.archivist.AbstractTest;
import com.zorroa.archivist.domain.*;
import com.zorroa.archivist.service.DyHierarchyService;
import com.zorroa.sdk.domain.PagedList;
import com.zorroa.sdk.domain.Pager;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;

import static junit.framework.TestCase.assertFalse;
import static junit.framework.TestCase.assertTrue;
import static org.junit.Assert.assertEquals;

/**
 * Created by chambers on 7/15/16.
 */
public class DyHierarchyDaoTests extends AbstractTest {

    @Autowired
    DyHierarchyService dyHierarchyService;

    @Autowired
    DyHierarchyDao dyHierarchyDao;

    DyHierarchy dyhi;
    Folder folder;

    @Before
    public void init() {
        folder = folderService.create(new FolderSpec("foo"), false);
        DyHierarchySpec spec = new DyHierarchySpec();
        spec.setFolderId(folder.getId());
        spec.setLevels(
                ImmutableList.of(
                        new DyHierarchyLevel("source.date", DyHierarchyLevelType.Day),
                        new DyHierarchyLevel("source.type.raw"),
                        new DyHierarchyLevel("source.extension.raw"),
                        new DyHierarchyLevel("source.filename.raw")));
        dyhi = dyHierarchyDao.create(spec);
    }

    @Test
    public void testCreate() {
        assertEquals(dyhi.getFolderId(), folder.getId());
        assertEquals(4, dyhi.getLevels().size());
    }

    @Test
    public void testGetByFolder() {
        DyHierarchy d2 = dyHierarchyDao.get(folder);
        assertEquals(dyhi, d2);
    }

    @Test
    public void testGet() {
        DyHierarchy d2 = dyHierarchyDao.get(dyhi.getId());
        assertEquals(dyhi, d2);
    }

    @Test
    public void testGetSetWorking() {
        assertFalse(dyHierarchyDao.isWorking(dyhi));
        assertTrue(dyHierarchyDao.setWorking(dyhi, true));
        assertTrue(dyHierarchyDao.isWorking(dyhi));
        assertFalse(dyHierarchyDao.setWorking(dyhi, true));
        assertTrue(dyHierarchyDao.isWorking(dyhi));
        assertTrue(dyHierarchyDao.setWorking(dyhi, false));
    }

    @Test
    public void testDelete() {
        assertTrue(dyHierarchyDao.delete(dyhi.getId()));
    }

    @Test
    public void testUpdate() {
        boolean result = dyHierarchyDao.update(dyhi.getId(),
                dyhi.setLevels(ImmutableList.of(new DyHierarchyLevel("source.filename.raw"))));
        assertTrue(result);
        DyHierarchy d2 = dyHierarchyDao.refresh(dyhi);
        assertEquals(1, d2.getLevels().size());
    }

    @Test
    public void testGetAll() {
        long count = dyHierarchyDao.count();
        List<DyHierarchy> list = dyHierarchyDao.getAll();
        assertEquals(count, list.size());

        Folder f2 = folderService.create(new FolderSpec("bar"), false);
        DyHierarchySpec spec = new DyHierarchySpec();
        spec.setFolderId(f2.getId());
        spec.setLevels(
                ImmutableList.of(
                        new DyHierarchyLevel("source.date", DyHierarchyLevelType.Day)));
        dyHierarchyDao.create(spec);
        list = dyHierarchyDao.getAll();
        assertEquals(count+1, list.size());
    }


    @Test
    public void testGetAllPaged() {
        long count = dyHierarchyDao.count();
        PagedList<DyHierarchy> list = dyHierarchyDao.getAll(Pager.first());
        assertEquals(count, list.size());

        Folder f2 = folderService.create(new FolderSpec("bar"), false);
        DyHierarchySpec spec = new DyHierarchySpec();
        spec.setFolderId(f2.getId());
        spec.setLevels(
                ImmutableList.of(
                        new DyHierarchyLevel("source.date", DyHierarchyLevelType.Day)));
        dyHierarchyDao.create(spec);
        list = dyHierarchyDao.getAll(Pager.first());
        assertEquals(count+1, list.size());
    }
}

File: archivist/src/test/kotlin/com/zorroa/archivist/repository/DyHierarchyDaoTests.kt
package com.zorroa.archivist.repository

import com.google.common.collect.ImmutableList
import com.zorroa.archivist.AbstractTest
import com.zorroa.archivist.domain.*
import org.junit.Before
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired

import junit.framework.TestCase.assertFalse
import junit.framework.TestCase.assertTrue
import org.junit.Assert.assertEquals

/**
 * Created by chambers on 7/15/16.
 */
class DyHierarchyDaoTests : AbstractTest() {

    @Autowired
    internal var dyHierarchyDao: DyHierarchyDao? = null

    lateinit var dyhi: DyHierarchy
    lateinit var folder: Folder

    @Before
    fun init() {
        folder = folderService.create(FolderSpec("foo"), false)
        val spec = DyHierarchySpec()
        spec.folderId = folder.id
        spec.levels = ImmutableList.of(
                DyHierarchyLevel("source.date", DyHierarchyLevelType.Day),
                DyHierarchyLevel("source.type.raw"),
                DyHierarchyLevel("source.extension.raw"),
                DyHierarchyLevel("source.filename.raw"))
        dyhi = dyHierarchyDao!!.create(spec)
    }

    @Test
    fun testCreate() {
        assertEquals(dyhi.folderId, folder.id)
        assertEquals(4, dyhi.levels.size.toLong())
    }

    @Test
    fun testGetByFolder() {
        val d2 = dyHierarchyDao!![folder]
        assertEquals(dyhi, d2)
    }

    @Test
    fun testGet() {
        val d2 = dyHierarchyDao!!.get(dyhi.id)
        assertEquals(dyhi, d2)
    }


    @Test
    fun testDelete() {
        assertTrue(dyHierarchyDao!!.delete(dyhi.id))
    }

    @Test
    fun testUpdate() {
        val result = dyHierarchyDao!!.update(dyhi.id,
                dyhi.setLevels(ImmutableList.of(DyHierarchyLevel("source.filename.raw"))))
        assertTrue(result)
        val d2 = dyHierarchyDao!!.refresh(dyhi)
        assertEquals(1, d2.levels.size.toLong())
    }

    @Test
    fun testGetAll() {
        val count = dyHierarchyDao!!.count()
        var list = dyHierarchyDao!!.getAll()
        assertEquals(count, list.size.toLong())

        val (id) = folderService.create(FolderSpec("bar"), false)
        val spec = DyHierarchySpec()
        spec.folderId = id
        spec.levels = ImmutableList.of(
                DyHierarchyLevel("source.date", DyHierarchyLevelType.Day))
        dyHierarchyDao!!.create(spec)
        list = dyHierarchyDao!!.getAll()
        assertEquals(count + 1, list.size.toLong())
    }


    @Test
    fun testGetAllPaged() {
        val count = dyHierarchyDao!!.count()
        var list = dyHierarchyDao!!.getAll(Pager.first())
        assertEquals(count, list.size().toLong())

        val (id) = folderService.create(FolderSpec("bar"), false)
        val spec = DyHierarchySpec()
        spec.folderId = id
        spec.levels = ImmutableList.of(
                DyHierarchyLevel("source.date", DyHierarchyLevelType.Day))
        dyHierarchyDao!!.create(spec)
        list = dyHierarchyDao!!.getAll(Pager.first())
        assertEquals(count + 1, list.size().toLong())
    }
}
